docs/ics/ics-xxx-signed-messages.md

```diff
@@ -0,0 +1,150 @@
+cryptographic algorithms such as having resistance to collision and second
+pre-image attacks, as well as being deterministic and uniform.
+
+### Encoding
+
+Our goal is to create a deterministic, injective, machine-verifiable means of
+encoding human-readable typed and structured data.
+
+Let us consider the set of signed messages to be: `B ∪ S`, where `B` is the set
+of byte strings and `S` is the set of human-readable typed structures. Thus, the
+set can can be encoded in a deterministic and injective way via the following
+rules, where `||` denotes concatenation:
+
+* `encode(b : B)` = `p1 || bytes("Signed Cosmos SDK Message: \n") || l || b`, where
+  * `l`: little endian uint64 encoding of the length of `b`
+  * `p1`: prefix to distinguish from normal transactions and other encoding cases
+* `encode(s : S, domainSeparator : B)` = `p2 || domainSeparator || amino(s)`, where
+  * `domainSeparator`: 32 byte encoding of the domain separator [see below](###DomainSeparator)
+  * `p2`: prefix to distinguish from normal transactions and other encoding cases
+
+> TODO: Figure out byte(s) prefix in the encoding to not have collisions with
+typical transaction signatures (JSON-encoded) and to distinguish the individual
+cases. This may require introducing prefixes to transactions.
+
+#### Assumptions
+
+`amino` is deterministic and seems to be injective -- at least when when
+concrete types are registered. Can we rely on this mechanism or do we need a
+custom true injective encoding?
+
+User-agents such Voyager, Metamask, and Ledger may provide human-readable and
+structured data, but this data must be consistent in terms of the field types
+and structure itself encoded via `amino`. This would require said applications
+to have an `amino` implementation. This may yet further warrant a simple custom
+encoding.
+
+> TODO: Do we need to implement a custom encoding scheme instead of or atop 
+amino similar or equal to EIP-712 `hashStruct`? As far as I know, this is mostly
+due to providing Soldity efficiencies and addressing limitations.
+
+### DomainSeparator
+
+Encoding structures can still lead to potential collisions and while this may be
+ok or even desired, it introduces a concern in that it could lead to two compatible
+signatures. The domain separator prevents collisions of otherwise identical
+structures. It is designed to unique per application use and is directly used in
+the signature encoding itself. The domain separator is also extensible where the
+protocol and application designer may introduce or omit fields to their liking,
+but we will provide a typical structure that can be used for proper separation
+of concerns:
+
+```golang
+type DomainSeparator struct {
+    name    string  // A user readable name of the signing origin or application.
+    chainID string  // The corresponding Cosmos chain identifier.
+    version uint16  // Version of the domain separator. A single major version should suffice.
```

Comment:
What is this field for / why would we version a `DomainSeparator`?

Reply:
(What would a new version add? Amino versioning would happen separately anyways.)

Reply:
This borrowed from EIP-712 where for example a domain separator could be update (omits old fields or adds new ones) for an application context. 

The domain separator can be omitted in our example, but again it's up to the protocol implementer.